### ISSUE TITLE
508-defect-1 ❗ Launchblocker [DATA SAFETY]: If an email address is provided, it MUST be validated

### ISSUE THREAD

**Issue by 1Copenut** [department-of-veterans-affairs/va.gov-team#12982]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

**Feedback framework**
- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Description

For the email address field, the re-enter field is intended to confirm the correct email address is entered. If an email is entered in the first field, the user **must** re-enter the email in a second field. The re-enter field should be a required field, and the user should be unable to continue. The entry in the first email address field should match the entry in the second field.

Due to the potential harm of accidentally sending personal information to a stranger, this is considered a 508-defect-1 launchblocker issue.

## Related Issues
* https://github.com/department-of-veterans-affairs/va.gov-team/issues/12585

## Point of Contact
**VFS Point of Contact:** _Trevor_

## Acceptance Criteria
- [ ] As a person entering a 686 dependent change, I want to be sure my email address is accurate and information is going to the intended destination

## Environment
* Operating System: all
* Browser: all
* Server destination: production

## Steps to Recreate
1. Enter https://staging.va.gov/view-change-dependents/add-remove-form-686c/ in browser
2. Navigate to the **Veteran's information** section
3. Enter an email address in the first email address input
4. Select Continue
5. Verify that it is possible to continue the form without validating the email address match

## Possible Fixes (optional)
- Recommend creating logic that the two email address input fields must match before a user may proceed.
- Consider pattern matching that the email entered is a well-formed email
- Provide useful, plain language error messages
    * If the second field is empty
    * If the email address is not a well-formed email

## WCAG or Vendor Guidance (optional)

Note: WUHCAG links provided additionally because they are generally more plain language, easier to understand.

* [W3C WCAG - Success Criterion 3.3.4 Error Prevention (Legal, Financial, Data) (Level AA):](https://www.w3.org/WAI/WCAG21/Understanding/error-prevention-legal-financial-data.html)
* [W3C WCAG - Error Prevention (Legal, Financial, Data): Understanding SC 3.3.4](https://www.w3.org/TR/UNDERSTANDING-WCAG20/minimize-error-reversible.html)
* [WUHCAG - 3.3.4 – Error Prevention (Legal, Financial, Data) (Level AA)](https://www.wuhcag.com/error-prevention-legal-financial-data/)
* [W3C WCAG - 3.3.6 Error Prevention (All) (Level AAA)](https://www.w3.org/TR/UNDERSTANDING-WCAG20/minimize-error-reversible-all.html)
* [WUHCAG - 3.3.6 – Error Prevention (All) (Level AAA)](https://www.wuhcag.com/error-prevention-all/)

## Screenshots or Trace Logs

![Screen Shot 2020-08-28 at 12 02 22 PM](https://user-images.githubusercontent.com/934879/91593995-c3217400-e926-11ea-9280-b92d832483fa.png)

**Comment by jason-gcio** [comment]
@1Copenut - the email confirmation is in place and I thought you might like to validate it.  (The first email field is coming in from the user object and is pre-filled.) User 228 results in this outcome:
![email-confirm](https://user-images.githubusercontent.com/53826255/92127094-f785ac00-edce-11ea-89a3-e4d802a8e647.JPG)

Nice job @micahchiang

**Comment by jason-gcio** [comment]
@1Copenut - Also, the field validation is also in place.  Is there anything else you can see that we should cover before closing this ticket?
![image](https://user-images.githubusercontent.com/53826255/92260890-96c9a280-eea6-11ea-9962-28041d7156ba.png)

**Comment by 1Copenut** [comment]
@jason-gcio I reviewed this on Staging just now. I like the implementation a lot and the checks it puts in place. Are there ways to prevent the second error message from firing if I skip over the confirm email address field, then come back to it? Screen reader users might have a difficult time with the current implementation, because the second error fires immediately after I start typing into the field. It would be better to check that the values match after the `onBlur()` event fires if possible. I'm attaching an animated GIF to illustrate the first error messaging firing, then the second error firing.

---
![confirm-email-error-handling](https://user-images.githubusercontent.com/934879/92271801-0d1dd300-eeae-11ea-8ee3-18fc3902a487.gif)

**Comment by micahchiang** [comment]
@1Copenut That's a good thought for sure. I can definitely investigate that issue. Just so we're on the same page - does the current implementation remove our launch blocker, or should we really try and see if firing the validation check after `onBlur()` is possible prior to launch?

**Comment by 1Copenut** [comment]
Great question @micahchiang ! If it's feasible to adjust the validation check for launch with low/no risk, let's go ahead and do that. If it's a bit more work, LMK and I'll open a fast follow ticket. Either way, this solves the immediate data safety issue, so should remove the launch block as is.

**Comment by jason-gcio** [comment]
@1Copenut - @micahchiang and I chatted and we decided to look into it and if it's quick and easy (famous last words) we will pop it in; else we will put this in the fast follow queue.  Either way, we will holler and make sure you are aware.
